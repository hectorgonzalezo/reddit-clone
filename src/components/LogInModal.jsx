import React, { useState, useRef, useEffect } from 'react';
import Button from './Button';
import googleIcon from '../assets/google.png';
import loadingIcon from '../assets/loading.gif';
import { func } from 'prop-types';
import { authorization } from '../firebase/firebase';

function LogInModal({ closeFunc }) {
  const formRef = useRef();
  const userNameRef = useRef();
  const passwordRef = useRef();
  const [usernameDoesntExist, setUsernameDoesntExist] = useState(false);
  const [loadingData, setLoadingData] = useState(false);

  async function submitLogIn(e) {
    e.preventDefault();
    const username = userNameRef.current.value;
    const password = passwordRef.current.value;
    try {
      // add loading animation to button
      setLoadingData(true);
      const account = await authorization.logIn(username, password);
      setUsernameDoesntExist(false);
      closeFunc();
    } catch (error) {
      setLoadingData(false);
      // If email already exists
      setUsernameDoesntExist(true);
    }
  }

  useEffect(() => {
    // add red outline to input fields if combination is invalid
    if (usernameDoesntExist) {
      userNameRef.current.classList.add('invalid');
      passwordRef.current.classList.add('invalid');
    } else {
      userNameRef.current.classList.remove('invalid');
      passwordRef.current.classList.remove('invalid');
    }
  }, [usernameDoesntExist]);

  return (
    <div id="logIn-outer">
      <div id="logIn-inner">
        <div>
          <button type="button" className="close-button" onClick={closeFunc}>
            x
          </button>
          <h1>Log In</h1>
          <p>
            By continuing, you are setting up a Reddit account and agree to our{" "}
            <a href="https://www.redditinc.com/policies/user-agreement">
              User Agreement
            </a>{" "}
            and{" "}
            <a href="https://www.reddit.com/policies/privacy-policy">
              Privacy Policy
            </a>
            .
          </p>
          <button onClick={authorization.logInPopup} className="google-button">
            <img src={googleIcon} alt="" />
            Continue with Google
          </button>
        </div>

        <div className="division">
          <hr />
          <h3>or</h3>
          <hr />
        </div>
        <form action="" ref={formRef}>
          <div className="input-wrap">
            <span />
            <input
              type="text"
              placeholder="0"
              name="username"
              id="username"
              ref={userNameRef}
              pattern="[A-Za-z0-9]*"
              maxLength={25}
              required
            />
            <label htmlFor="username">Username</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="password"
              id="password"
              ref={passwordRef}
              minLength={6}
              autoComplete="on"
              required
            />
            <label htmlFor="password">Password</label>
          </div>
          {usernameDoesntExist ? <span>Incorrect username or password</span> : null}
          <Button text="" type="submit" onClick={submitLogIn} >
          {loadingData ? <img src={loadingIcon} alt="loading" data-testid="loading-icon" /> : ['Log In']}
          </Button>
        </form>
        <p>
          First time in reddit? <a href="">Sign Up</a>
        </p>
      </div>
    </div>
  );
}

LogInModal.defaultProps = {
  closeFunc: () => {},
};

LogInModal.propTypes = {
  closeFunc: func,
};

export default LogInModal;
