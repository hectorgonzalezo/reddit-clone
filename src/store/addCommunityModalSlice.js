import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state) => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
