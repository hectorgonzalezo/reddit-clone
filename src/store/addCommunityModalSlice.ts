import { createSlice } from '@reduxjs/toolkit';

export const addCommunityModalSlice = createSlice({
  name: 'addCommunityModalVisible',
  initialState: false,
  reducers: {
    toggleAddCommunityModal: (state) => !state
  },
});

export const { toggleAddCommunityModal } = addCommunityModalSlice.actions;

export const selectAddCommunityModalVisibility = (state: AddCommunityModalState): boolean => state.addCommunityModalVisible;

export default addCommunityModalSlice.reducer;
