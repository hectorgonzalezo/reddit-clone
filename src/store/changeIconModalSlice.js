import { createSlice } from '@reduxjs/toolkit';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => {
      return !state
    }
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state) => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
