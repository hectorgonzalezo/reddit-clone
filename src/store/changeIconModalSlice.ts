import { createSlice } from '@reduxjs/toolkit';
import { ChangeIconModalState } from '../type';

export const changeIconModalSlice = createSlice({
  name: 'changeIconModalVisible',
  initialState: false,
  reducers: {
    toggleChangeIconModal: (state, action) => !state
  },
});

export const { toggleChangeIconModal } = changeIconModalSlice.actions;

export const selectChangeIconModalVisibility = (state: ChangeIconModalState): boolean => state.changeIconModalVisible;

export default changeIconModalSlice.reducer;
