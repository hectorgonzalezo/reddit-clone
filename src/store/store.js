import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
