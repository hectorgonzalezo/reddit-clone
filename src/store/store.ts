import { configureStore } from '@reduxjs/toolkit';
import userReducer from './userSlice';
import postsReducer from './postsSlice';
import subredditsReducer from './subredditsSlice';
import loginModalReducer from './loginModalSlice';
import changeIconModalReducer from './changeIconModalSlice';
import addCommunityModalReducer from './addCommunityModalSlice';
import currentSubredditReducer from './currentSubredditSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    posts: postsReducer,
    subreddits: subredditsReducer,
    loginModalVisible: loginModalReducer,
    changeIconModalVisible: changeIconModalReducer,
    addCommunityModalVisible: addCommunityModalReducer,
    currentSubreddit: currentSubredditReducer,
  }
});

export default store;
