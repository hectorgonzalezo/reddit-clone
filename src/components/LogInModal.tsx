import React, { useState, useRef, useEffect, SyntheticEvent } from 'react';
import { useDispatch } from 'react-redux';
import { logIn } from '../api/users';
import { addUser } from '../store/userSlice';
import Button from './Button';
import loadingIcon from '../assets/loading.gif';

interface LogInModalProps {
  closeFunc: () => void;
};

function LogInModal({ closeFunc= () => {} }: LogInModalProps): JSX.Element {
  const formRef = useRef<HTMLFormElement>(null);
  const userNameRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const [usernameDoesntExist, setUsernameDoesntExist] = useState(false);
  const [loadingData, setLoadingData] = useState(false);
  const dispatch = useDispatch();

  async function submitLogIn(e: SyntheticEvent): Promise<void> {
    e.preventDefault();
    if (userNameRef.current !== null && passwordRef.current !== null){
      const username = userNameRef.current.value;
      const password = passwordRef.current.value;
    try {
      setLoadingData(true);
      const fetchedUser = await logIn({username, password});
      // add loading animation to button
      setUsernameDoesntExist(false);
      // update redux store
      dispatch(addUser(fetchedUser));
      closeFunc();
    } catch (error) {
      setLoadingData(false);
      // If email already exists
      setUsernameDoesntExist(true);
    }
  }
  }

  useEffect(() => {
    // add red outline to input fields if combination is invalid
    if (userNameRef.current !== null && passwordRef.current !== null) {
      if (usernameDoesntExist) {
        userNameRef.current.classList.add("invalid");
        passwordRef.current.classList.add("invalid");
      } else {
        userNameRef.current.classList.remove("invalid");
        passwordRef.current.classList.remove("invalid");
      }
    }
  }, [usernameDoesntExist]);

  return (
    <div id="logIn-outer" className="modal-outer">
      <div id="logIn-inner" className="modal-inner">
        <div>
          <button type="button" className="close-button" onClick={closeFunc} aria-label="close menu">
            x
          </button>
          <h1>Log In</h1>
          <p>
            By continuing, you are setting up a Reddit account and agree to our{" "}
            <a href="https://www.redditinc.com/policies/user-agreement">
              User Agreement
            </a>{" "}
            and{" "}
            <a href="https://www.reddit.com/policies/privacy-policy">
              Privacy Policy
            </a>
            .
          </p>
        </div>
        <div className="division"/>
        <form action="" ref={formRef}>
          <div className="input-wrap">
            <span />
            <input
              type="text"
              placeholder="0"
              name="username"
              id="username"
              ref={userNameRef}
              pattern="[A-Za-z0-9]*"
              maxLength={25}
              required
            />
            <label htmlFor="username">Username</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="password"
              id="password"
              ref={passwordRef}
              minLength={6}
              autoComplete="on"
              required
            />
            <label htmlFor="password">Password</label>
          </div>
          {usernameDoesntExist ? (
            <span>Incorrect username or password</span>
          ) : null}
          <Button text="" type="submit" onClick={submitLogIn}>
            {loadingData ? (
              <img
                src={loadingIcon}
                alt="loading"
                data-testid="loading-icon"
                className="loading-icon"
              />
            ) : (
              ["Log In"]
            )}
          </Button>
        </form>
        <p>
          First time in reddit? <a href="">Sign Up</a>
        </p>
      </div>
    </div>
  );
}



export default LogInModal;
