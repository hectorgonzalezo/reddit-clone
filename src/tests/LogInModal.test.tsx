import React from 'react';
import { render, screen, container, act } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import LogInModal from '../components/LogInModal';
import store from '../store/store';

jest.mock('../api/users');

describe('Styled button works', () => {
  test('Filling in with fake data shows error message', async () => {
    render(<Provider store={store}>
            <LogInModal />
           </Provider>
    );
    const textInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);
    const button = screen.getByRole('button', { name: 'Log In' });

    userEvent.type(textInput, 'fakefake');
    userEvent.type(passwordInput, 'fakefake');
    await act(async () => userEvent.click(button));

    expect(screen.getByText('Incorrect username or password')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('Pressing log in button shows loading gif', async () => {
    render(<Provider store={store}>
      <LogInModal />
     </Provider>
    );
    const textInput = screen.getByRole('textbox');
    const passwordInput = screen.getByLabelText(/password/i);
    const button = screen.getByRole('button', { name: 'Log In' });

    userEvent.type(textInput, 'truetrue');
    userEvent.type(passwordInput, 'truetrue');
    await act(async() => userEvent.click(button));

    expect(screen.getByTestId('loading-icon')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});