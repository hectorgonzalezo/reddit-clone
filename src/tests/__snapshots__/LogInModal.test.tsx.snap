// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled button works Filling in with fake data shows error message 1`] = `undefined`;

exports[`Styled button works Pressing log in button shows loading gif 1`] = `undefined`;
